### PR TITLE
Make provided fork work in takeEvery and takeLatest workers.

### PR DESCRIPTION
**Pull request by jessjenk**
I included a couple tests that illustrate the problem I was having. I'm not sure if this is the best fix, but it's works for both the tests and my somewhat-more-complicated use case in my project.

### CODE REVIEW

**Comment by jessjenk** [comment]
I see the CI test is failing, I think because of a lint error in a part of the test file that I didn't change. I'm happy to fix it if you want, though!

**Comment by jfairbank** [comment]
Hi, @jessjenk!

Thanks for catching this and opening a PR.

I think you've found a general issue with nested forks and providers. After looking over the issue you found in your test case, I discovered the general problem with this test:

```js
test('provides fork values in nested forks', () => {
  const fakeTask = { hello: 'world' };

  function* otherSaga() {
    yield 42;
  }

  function* fooForkSaga(arg1, arg2) {
    const task = yield fork(otherSaga);
    yield put({
      type: 'FORKED_TASK',
      payload: task,
      meta: { args: [arg1, arg2] },
    });
  }

  function* saga() {
    yield fork(fooForkSaga, 42, 'hello');
  }

  return expectSaga(saga)
    .provide({
      fork({ fn }, next) {
        if (fn === otherSaga) {
          return fakeTask;
        }

        return next();
      },
    })
    .put({
      type: 'FORKED_TASK',
      payload: fakeTask,
      meta: {
        args: [42, 'hello'],
      },
    })
    .dispatch({ type: 'REQUEST_USER' })
    .run({ silenceTimeout: true });
});
```

Your fix handles this when the top level fork happens to be a helper `takeEvery`/`takeLatest` because it ensures we process the `takeEvery`/`takeLatest` fork as normal with redux-saga. Then, we get the opportunity to provide the fake task for the nested fork later.

Unfortunately, this won't fix the more general case like the test I added above. However, I _think_ we can fix this in general with something like this instead.

```js
      case type === FORK: {
        const { args, detached, context, fn } = effect;
        const yieldedHelperEffect = isHelper(fn);

        const isProvided = useProvidedValue(value) !== value;

        if (!detached && !isProvided) {
```

Basically, we check to see if this particular fork will be mocked out with a provided value if the return value doesn't match the current effect. It's not ideal because it forces us to call the fork provider twice for every fork effect, though.

Could you try this fix out to see if it works for your use case? Also, I bet we'll have a similar issue with the spawn effect. Would you mind writing test cases for it as well please? Also, I'm open to other general fixes for this if you have any ideas.

Thanks so much for opening the PR again!

**Comment by jessjenk** [comment]
@jfairbank Thanks for the feedback! Your suggested fix makes sense to me. I modified it a bit to avoid calling `useProvidedValue` as much, applied it to `spawn` as well, and added tests for spawn.

**Comment by jfairbank** [comment]
This is fantastic! Thanks so much. And good catch preventing the second call to `useProvidedValue`.
